File: TP2/Q09/QuickSort.c
#include <stdio.h>

// Função para trocar dois elementos de um array
void trocar(int* a, int * b){
    int aux = *a;
    *a = *b;
    *b = aux;
}
int OrganizarQuick(int esq, int dir, int* arr) {
    int Pivot = arr[(esq + dir) / 2]; // Valor do pivô, não o índice

    while (esq <= dir) {
        while (arr[esq] < Pivot) {//percorre ate achar um valor maior que o pivot
            esq++;
        }
        while (arr[dir] > Pivot) { //percorre ate achar um valor menor que o pivot
            dir--;
        }
        if (esq <= dir) { //se o valor da esquerda for menor que o da direita, troca
            trocar(&arr[esq], &arr[dir]);
            esq++;
            dir--;
        }
    }

    return esq; // Retorna o índice do pivô
}

void OrdenarQuick(int esq, int dir, int* arr){
    if(esq < dir){
        int pivo = OrganizarQuick(esq, dir, arr);
        OrdenarQuick(esq, pivo - 1, arr);
        OrdenarQuick(pivo + 1, dir, arr);
    }
}

int main() {
    int arr[] = {3, 6, 8, 10, 1, 2, 1};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("Array original:\n");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }

    quicksort(arr, 0, n - 1);

    printf("\nArray ordenado:\n");
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }

    return 0;
}

File: Ordenação/QuickSort.java
public class QuickSort {
    
    public static void QuickSort(int array[], int esq, int dir){

        int pivo = array[(esq + dir) / 2];
        int i = esq;
        int j = dir;
        while(i<j){
            while(array[i] < pivo){
                i++;
            }

            while(array[j]>pivo){
                j--;
            }
               int aux = array[i];
                array[i] = array[j];
                array[j] = aux;
                i++;j--;
            
        }

        if(i<dir){
            QuickSort(array, i, dir);
        }

        if(j>esq){
            QuickSort(array, esq, j);
        }
    }



    public static void main(String[] args) {
        int[] array = { 5, 3, 2, 4, 7, 1, 0, 6 };
        QuickSort(array, 0, array.length - 1);
        for (int elemento : array) {
            System.out.print(elemento + " ");
        }
    }



}
